src/typecheck/arith.ml

```diff
@@ -2,6 +2,95 @@
 module Id = Dolmen.Std.Id
 module Term = Dolmen.Std.Term
 
+(* Ae arithmetic *)
+(* ************************************************************************ *)
+module Ae = struct
+
+  module Tff
+      (Type : Tff_intf.S)
+      (Ty : Dolmen.Intf.Ty.Ae_Arith with type t := Type.Ty.t)
+      (T : Dolmen.Intf.Term.Ae_Arith with type t := Type.T.t
+                                        and type ty := Type.Ty.t) = struct
+
+      type _ Type.err +=
+      | Expected_arith_type : Type.Ty.t -> Term.t Type.err
+      | Incompatible_Arith_Types : Type.Ty.t * Type.Ty.t -> Term.t Type.err
+
+      let dispatch1 env (mk_int, mk_real) ast t =
+        let ty = T.ty t in
+        match Ty.view ty with
+        | `Int -> mk_int t
+        | `Real -> mk_real t
+        | _ -> Type._error env (Ast ast) (Expected_arith_type ty)
+
+      let dispatch2 ?(strict = true) env (mk_int, mk_real) ast a b =
+        let tya = T.ty a in
+        let tyb = T.ty b in
+        match Ty.view tya, Ty.view tyb with
```

Comment:
Assuming the `~strict` parameter is not actually used, then you could delete the `Incompatible arith types` error, and this match could be simplified greatly, by simply pattern matching on the type of the first argument, and applying the corresponding `mk_*` function. In that case, when the second argument has a different (but still arith) type, the error would be caught by the generic mechanism for when an argument does not have the expected type. However, it might result in slightly less precise error messages.

Reply:
Done.